spec/message.abnf

```diff
@@ -0,0 +1,57 @@
+message = [s] *(declaration [s]) body [s]
+
+declaration = let s variable [s] "=" [s] "{" [s] expression [s] "}"
+body = pattern
+     / (selectors 1*([s] variant))
+
+pattern = "{" *(text / placeholder) "}"
+selectors = match 1*([s] selector)
+selector = "{" [s] expression [s] "}"
+variant = when 1*(s key) [s] pattern
+key = nmtoken / literal / "*"
+
+placeholder = "{" [s] (expression / markup / markup-end) [s] "}"
```

Comment:
I think denormalization would help with readability here.
```suggestion
placeholder = "{" [s] expression [s] "}"
            / "{" [s] markup-start *(s option) [s] "}"
            / "{" [s] markup-end [s] "}"
```

And given the reuse of `"{" [s] expression [s] "}"`, it might make sense for that to have a name of its own.